File: server/src/test/java/com/rocketmq/studio/instance/InstanceServiceTest.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.rocketmq.studio.instance;

import com.rocketmq.studio.common.domain.enums.InstanceType;
import com.rocketmq.studio.common.exception.BusinessException;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.List;
import java.util.Optional;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
class InstanceServiceTest {

    @Mock
    private InstanceRepository instanceRepository;

    @InjectMocks
    private InstanceService instanceService;

    @Test
    void listInstancesShouldReturnAllWhenNoFilters() {
        List<InstanceVO> instances = List.of(
                InstanceVO.builder().name("inst-1").build(),
                InstanceVO.builder().name("inst-2").build()
        );
        when(instanceRepository.findAll()).thenReturn(instances);

        List<InstanceVO> result = instanceService.listInstances(null, null);

        assertThat(result).hasSize(2);
        verify(instanceRepository).findAll();
    }

    @Test
    void listInstancesShouldFilterByType() {
        List<InstanceVO> instances = List.of(
                InstanceVO.builder().name("proxy-1").type(InstanceType.PROXY).build()
        );
        when(instanceRepository.findByType(InstanceType.PROXY)).thenReturn(instances);

        List<InstanceVO> result = instanceService.listInstances(InstanceType.PROXY, null);

        assertThat(result).hasSize(1);
        verify(instanceRepository).findByType(InstanceType.PROXY);
    }

    @Test
    void listInstancesShouldSearchByKeyword() {
        List<InstanceVO> instances = List.of(
                InstanceVO.builder().name("production").build()
        );
        when(instanceRepository.search("prod")).thenReturn(instances);

        List<InstanceVO> result = instanceService.listInstances(null, "prod");

        assertThat(result).hasSize(1);
        verify(instanceRepository).search("prod");
    }

    @Test
    void listInstancesShouldTrimSearchKeyword() {
        List<InstanceVO> instances = List.of(
                InstanceVO.builder().name("production").build()
        );
        when(instanceRepository.search("prod")).thenReturn(instances);

        List<InstanceVO> result = instanceService.listInstances(null, " prod ");

        assertThat(result).hasSize(1);
        verify(instanceRepository).search("prod");
    }

    @Test
    void listInstancesShouldFilterByTypeAndSearch() {
        List<InstanceVO> instances = List.of(
                InstanceVO.builder().name("production-proxy").type(InstanceType.PROXY).build()
        );
        when(instanceRepository.findByTypeAndSearch(InstanceType.PROXY, "prod")).thenReturn(instances);

        List<InstanceVO> result = instanceService.listInstances(InstanceType.PROXY, "prod");

        assertThat(result).hasSize(1);
        verify(instanceRepository).findByTypeAndSearch(InstanceType.PROXY, "prod");
    }

    @Test
    void listInstancesShouldTrimSearchKeywordWhenFilteringByType() {
        List<InstanceVO> instances = List.of(
                InstanceVO.builder().name("production-proxy").type(InstanceType.PROXY).build()
        );
        when(instanceRepository.findByTypeAndSearch(InstanceType.PROXY, "prod")).thenReturn(instances);

        List<InstanceVO> result = instanceService.listInstances(InstanceType.PROXY, " prod ");

        assertThat(result).hasSize(1);
        verify(instanceRepository).findByTypeAndSearch(InstanceType.PROXY, "prod");
    }

    @Test
    void listInstancesShouldIgnoreBlankSearch() {
        List<InstanceVO> instances = List.of(InstanceVO.builder().name("inst").build());
        when(instanceRepository.findAll()).thenReturn(instances);

        List<InstanceVO> result = instanceService.listInstances(null, "   ");

        assertThat(result).hasSize(1);
        verify(instanceRepository).findAll();
    }

    @Test
    void createInstanceShouldSetIdAndTimestamps() {
        InstanceVO input = InstanceVO.builder()
                .name("new-instance")
                .endpoint("10.0.1.1:8080")
                .type(InstanceType.PROXY)
                .build();

        when(instanceRepository.save(any(InstanceVO.class))).thenAnswer(inv -> inv.getArgument(0));

        InstanceVO result = instanceService.createInstance(input);

        assertThat(result.getId()).isNotBlank();
        assertThat(result.getCreatedAt()).isNotNull();
        assertThat(result.getUpdatedAt()).isNotNull();
        assertThat(result.getName()).isEqualTo("new-instance");
        verify(instanceRepository).save(any(InstanceVO.class));
    }

    @Test
    void createInstanceShouldThrowWhenNameIsNull() {
        InstanceVO input = InstanceVO.builder()
                .name(null)
                .endpoint("10.0.1.1:8080")
                .build();

        assertThatThrownBy(() -> instanceService.createInstance(input))
                .isInstanceOf(BusinessException.class)
                .hasMessage("InstanceVO name is required");
    }

    @Test
    void createInstanceShouldThrowWhenNameIsBlank() {
        InstanceVO input = InstanceVO.builder()
                .name("  ")
                .endpoint("10.0.1.1:8080")
                .build();

        assertThatThrownBy(() -> instanceService.createInstance(input))
                .isInstanceOf(BusinessException.class)
                .hasMessage("InstanceVO name is required");
    }

    @Test
    void createInstanceShouldThrowWhenEndpointIsNull() {
        InstanceVO input = InstanceVO.builder()
                .name("valid-name")
                .endpoint(null)
                .build();

        assertThatThrownBy(() -> instanceService.createInstance(input))
                .isInstanceOf(BusinessException.class)
                .hasMessage("InstanceVO endpoint is required");
    }

    @Test
    void createInstanceShouldThrowWhenEndpointIsBlank() {
        InstanceVO input = InstanceVO.builder()
                .name("valid-name")
                .endpoint("  ")
                .build();

        assertThatThrownBy(() -> instanceService.createInstance(input))
                .isInstanceOf(BusinessException.class)
                .hasMessage("InstanceVO endpoint is required");
    }

    @Test
    void updateInstanceShouldMergeFieldsOntoExisting() {
        InstanceVO existing = InstanceVO.builder()
                .name("old-name")
                .endpoint("10.0.1.1:8080")
                .type(InstanceType.PROXY)
                .remark("old remark")
                .build();
        existing.setId("inst-1");

        InstanceVO update = InstanceVO.builder()
                .name("new-name")
                .remark("new remark")
                .build();
        update.setId("inst-1");

        when(instanceRepository.findById("inst-1")).thenReturn(Optional.of(existing));
        when(instanceRepository.save(any(InstanceVO.class))).thenAnswer(inv -> inv.getArgument(0));

        InstanceVO result = instanceService.updateInstance(update);

        assertThat(result.getName()).isEqualTo("new-name");
        assertThat(result.getEndpoint()).isEqualTo("10.0.1.1:8080");
        assertThat(result.getType()).isEqualTo(InstanceType.PROXY);
        assertThat(result.getRemark()).isEqualTo("new remark");
        assertThat(result.getUpdatedAt()).isNotNull();
    }

    @Test
    void updateInstanceShouldThrowWhenIdIsNull() {
        InstanceVO input = InstanceVO.builder().name("test").build();
        input.setId(null);

        assertThatThrownBy(() -> instanceService.updateInstance(input))
                .isInstanceOf(BusinessException.class)
                .hasMessage("InstanceVO ID is required");
    }

    @Test
    void updateInstanceShouldThrowWhenIdIsBlank() {
        InstanceVO input = InstanceVO.builder().name("test").build();
        input.setId("  ");

        assertThatThrownBy(() -> instanceService.updateInstance(input))
                .isInstanceOf(BusinessException.class)
                .hasMessage("InstanceVO ID is required");
    }

    @Test
    void updateInstanceShouldThrowWhenInstanceNotFound() {
        InstanceVO input = InstanceVO.builder().name("test").build();
        input.setId("nonexistent");

        when(instanceRepository.findById("nonexistent")).thenReturn(Optional.empty());

        assertThatThrownBy(() -> instanceService.updateInstance(input))
                .isInstanceOf(BusinessException.class)
                .hasMessage("InstanceVO not found: nonexistent");
    }

    @Test
    void deleteInstanceShouldRemoveExistingInstance() {
        InstanceVO existing = InstanceVO.builder().name("to-delete").build();
        existing.setId("inst-1");

        when(instanceRepository.findById("inst-1")).thenReturn(Optional.of(existing));

        instanceService.deleteInstance("inst-1");

        verify(instanceRepository).deleteById("inst-1");
    }

    @Test
    void deleteInstanceShouldThrowWhenIdIsNull() {
        assertThatThrownBy(() -> instanceService.deleteInstance(null))
                .isInstanceOf(BusinessException.class)
                .hasMessage("InstanceVO ID is required");
    }

    @Test
    void deleteInstanceShouldThrowWhenIdIsBlank() {
        assertThatThrownBy(() -> instanceService.deleteInstance("   "))
                .isInstanceOf(BusinessException.class)
                .hasMessage("InstanceVO ID is required");
    }

    @Test
    void deleteInstanceShouldThrowWhenInstanceNotFound() {
        when(instanceRepository.findById("missing")).thenReturn(Optional.empty());

        assertThatThrownBy(() -> instanceService.deleteInstance("missing"))
                .isInstanceOf(BusinessException.class)
                .hasMessage("InstanceVO not found: missing");
    }
}

File: web/src/services/instanceService.test.ts
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { beforeEach, describe, expect, it, vi } from 'vitest';

const dataModeMock = vi.hoisted(() => ({ isMockMode: vi.fn(() => true) }));
vi.mock('./dataMode', () => dataModeMock);
const instanceApiMock = vi.hoisted(() => ({
  listInstances: vi.fn(),
  createInstance: vi.fn(),
  updateInstance: vi.fn(),
  deleteInstance: vi.fn(),
  deleteInstancesBatch: vi.fn(),
}));
vi.mock('../api/instance', () => instanceApiMock);
vi.mock('../config', () => ({
  API_BASE_URL: '/api',
}));

import {
  createInstance,
  deleteInstance,
  deleteInstancesBatch,
  getInstanceCapabilities,
  listInstances,
  updateInstance,
} from './instanceService';
import type { Instance } from '../api/instance';

describe('instanceService mock instances', () => {
  it('returns defensive copies from list reads', async () => {
    const instances = await listInstances();
    const originalName = instances[0].name;
    const originalRemark = instances[0].remark;

    instances[0].name = 'mutated-name';
    instances[0].remark = 'mutated-remark';

    const fresh = await listInstances();

    expect(fresh[0].name).toBe(originalName);
    expect(fresh[0].remark).toBe(originalRemark);
    expect(fresh[0]).not.toBe(instances[0]);
  });

  it('filters mock instances with the same type and search semantics as the API', async () => {
    const byType = await listInstances({ type: 'DIRECT' });
    expect(byType.map((instance) => instance.id)).toEqual([1, 2]);

    const byEndpoint = await listInstances({ search: '  10.0.2.21  ' });
    expect(byEndpoint.map((instance) => instance.id)).toEqual([3]);

    const combined = await listInstances({ type: 'DIRECT', search: 'instance-direct-2' });
    expect(combined.map((instance) => instance.id)).toEqual([2]);

    const clusterOnly = await listInstances({ type: 'PROXY_CLUSTER' });
    expect(clusterOnly.map((instance) => instance.id)).toEqual([3, 4]);
    await expect(listInstances({ type: 'PROXY_LOCAL' })).resolves.toEqual([
      expect.objectContaining({ type: 'PROXY_LOCAL' }),
    ]);
    await expect(listInstances({ type: 'CLOUD' })).resolves.toEqual([]);
  });

  it('does not expose created or updated store records by reference', async () => {
    const created = await createInstance({
      name: 'rocketmq-copy-test',
      type: 'PROXY_CLUSTER',
      endpoint: 'proxy-copy-test:8080',
      remark: 'created',
    });

    created.name = 'mutated-created';
    created.remark = 'mutated-created-remark';

    const afterCreate = await listInstances();
    const storedCreated = afterCreate.find((instance) => instance.id === created.id);
    expect(storedCreated).toMatchObject({
      name: 'rocketmq-copy-test',
      type: 'PROXY_CLUSTER',
      remark: 'created',
    });

    const updated = await updateInstance({
      instanceId: 'rocketmq-copy-test',
      remark: 'updated',
    });
    updated.remark = 'mutated-updated';

    const afterUpdate = await listInstances();
    const storedUpdated = afterUpdate.find((instance) => instance.id === created.id);
    expect(storedUpdated?.remark).toBe('updated');
  });

  it('rejects deleting missing mock instances', async () => {
    const before = await listInstances();

    await expect(deleteInstance('missing-instance')).rejects.toThrow(
      'Instance not found: missing-instance',
    );

    await expect(listInstances()).resolves.toEqual(before);
  });

  it('returns provider-specific mock capabilities without sharing mutable arrays', async () => {
    const first = await getInstanceCapabilities('instance-direct-1');
    expect(first.instanceId).toBe('instance-direct-1');
    expect(first.vendor).toBe('APACHE');
    first.capabilities.length = 0;

    const second = await getInstanceCapabilities('instance-direct-1');

    expect(second.capabilities).toEqual(
      expect.arrayContaining(['DLQ_MANAGEMENT', 'DIRECT_MESSAGE_CONSUME']),
    );
    await expect(getInstanceCapabilities('missing-instance')).rejects.toThrow(
      'Instance not found: missing-instance',
    );
  });
});

describe('instanceService list request dedupe', () => {
  beforeEach(() => {
    dataModeMock.isMockMode.mockReturnValue(true);
    instanceApiMock.listInstances.mockReset();
  });

  it('shares one inflight list request between concurrent callers', async () => {
    dataModeMock.isMockMode.mockReturnValue(false);
    const fixture: Instance[] = [
      {
        id: 1,
        name: 'shared-instance',
        remark: null,
        type: 'PROXY_CLUSTER',
        endpoint: '10.0.0.1:8080',
        topicCount: 0,
        consumerGroupCount: 0,
        gmtCreate: '2026-01-01T00:00:00Z',
        gmtModified: '2026-01-01T00:00:00Z',
      },
    ];
    let resolveList!: (value: Instance[]) => void;
    instanceApiMock.listInstances.mockImplementation(
      () =>
        new Promise<Instance[]>((resolve) => {
          resolveList = resolve;
        }),
    );

    const first = listInstances({});
    const second = listInstances({ search: '   ' });
    resolveList(fixture);

    const [a, b] = await Promise.all([first, second]);
    expect(instanceApiMock.listInstances).toHaveBeenCalledTimes(1);
    expect(a).toEqual(b);
    expect(a).not.toBe(b);

    instanceApiMock.listInstances.mockResolvedValue(fixture);
    await listInstances({});
    expect(instanceApiMock.listInstances).toHaveBeenCalledTimes(2);
  });

  it('does not share inflight requests across data modes', async () => {
    dataModeMock.isMockMode.mockReturnValue(false);
    let resolveRealList!: (value: Instance[]) => void;
    instanceApiMock.listInstances.mockImplementationOnce(
      () =>
        new Promise<Instance[]>((resolve) => {
          resolveRealList = resolve;
        }),
    );

    const realRequest = listInstances({});
    dataModeMock.isMockMode.mockReturnValue(true);
    const mockResult = await listInstances({});

    expect(mockResult.length).toBeGreaterThan(0);
    expect(instanceApiMock.listInstances).toHaveBeenCalledTimes(1);

    resolveRealList([]);
    await expect(realRequest).resolves.toEqual([]);
  });
});

describe('instanceService dedupe invalidation after mutations', () => {
  beforeEach(() => {
    dataModeMock.isMockMode.mockReturnValue(false);
    instanceApiMock.listInstances.mockReset();
  });

  function instanceFixture(name: string, remark: string | null = null): Instance {
    return {
      id: 1,
      name,
      remark,
      type: 'PROXY_CLUSTER',
      endpoint: '10.0.0.1:8080',
      topicCount: 0,
      consumerGroupCount: 0,
      gmtCreate: '2026-01-01T00:00:00Z',
      gmtModified: '2026-01-01T00:00:00Z',
    };
  }

  function holdFirstListRequest(): { resolve: (value: Instance[]) => void } {
    const holder: { resolve?: (value: Instance[]) => void } = {};
    instanceApiMock.listInstances.mockImplementationOnce(
      () =>
        new Promise<Instance[]>((resolve) => {
          holder.resolve = resolve;
        }),
    );
    return holder as { resolve: (value: Instance[]) => void };
  }

  it('does not serve the pre-create snapshot to a list request issued after createInstance', async () => {
    const before = [instanceFixture('kept')];
    const after = [instanceFixture('kept'), instanceFixture('created')];
    const first = holdFirstListRequest();
    instanceApiMock.listInstances.mockResolvedValueOnce(after);

    const staleRead = listInstances({});
    await createInstance({
      name: 'created',
      type: 'PROXY_CLUSTER',
      endpoint: '10.0.0.2:8080',
    });
    const refresh = listInstances({});
    first.resolve(before);

    await expect(staleRead).resolves.toEqual(before);
    await expect(refresh).resolves.toEqual(after);
    expect(instanceApiMock.listInstances).toHaveBeenCalledTimes(2);
  });

  it('does not serve the pre-update snapshot to a list request issued after updateInstance', async () => {
    const before = [instanceFixture('kept', 'old')];
    const after = [instanceFixture('kept', 'updated')];
    const first = holdFirstListRequest();
    instanceApiMock.listInstances.mockResolvedValueOnce(after);

    const staleRead = listInstances({});
    await updateInstance({ instanceId: 'kept', remark: 'updated' });
    const refresh = listInstances({});
    first.resolve(before);

    await expect(staleRead).resolves.toEqual(before);
    await expect(refresh).resolves.toEqual(after);
    expect(instanceApiMock.listInstances).toHaveBeenCalledTimes(2);
  });

  it('does not serve the pre-delete snapshot to a list request issued after deleteInstance', async () => {
    const before = [instanceFixture('removed'), instanceFixture('kept')];
    const after = [instanceFixture('kept')];
    const first = holdFirstListRequest();
    instanceApiMock.listInstances.mockResolvedValueOnce(after);

    const staleRead = listInstances({});
    await deleteInstance('removed');
    const refresh = listInstances({});
    first.resolve(before);

    await expect(staleRead).resolves.toEqual(before);
    await expect(refresh).resolves.toEqual(after);
    expect(instanceApiMock.listInstances).toHaveBeenCalledTimes(2);
  });

  it('does not serve the pre-delete snapshot to a list request issued after deleteInstancesBatch', async () => {
    const before = [instanceFixture('removed'), instanceFixture('kept')];
    const after = [instanceFixture('kept')];
    const first = holdFirstListRequest();
    instanceApiMock.listInstances.mockResolvedValueOnce(after);

    const staleRead = listInstances({});
    await deleteInstancesBatch(['removed']);
    const refresh = listInstances({});
    first.resolve(before);

    await expect(staleRead).resolves.toEqual(before);
    await expect(refresh).resolves.toEqual(after);
    expect(instanceApiMock.listInstances).toHaveBeenCalledTimes(2);
  });
});
